### PR TITLE
Upload a database dump s3

### PR DESCRIPTION
**Pull request by eddierubeiz**
Ref #934 .

This creates a new file on S3, opens a write stream to it, and directly dumps the database into the write stream; the dump is not written to a temp file.

The dump is a logical (not physical) backup of the database (i.e.: not coupled to a particular database version; slower to restore; more portable;  consistent with pre-Heroku backup workflow; encoded as a UTF-8 file rather than in binary.)

Note that Heroku also offers physical backups which are fast and useful for short-term DB maneuvers. This workflow is meant to supplement, not replace, those backups.

We create a write stream into a new s3 object, then `pg_dump` the database directly into the write stream. No temp file is saved to disk.

### CODE REVIEW

**Comment by jrochkind** [comment]
Looking good!

* Is `'PGSql/digcol_backup.dump` the current location of the thing? 

* I am not totally sure what you mean by "Consider uploading an ASCII SQL file rather than a binary .dump file." -- what was our previous backup routine doing? If different from what this is doing, is there any way to get heroku to do the thing we used to do be doing? If not, what are the implications of doing a binary dump instead?

* If we want to use different AWS credentials for backup, so that we don't have the AWS credentials normally used by our app with access to backup location, which is I think what we want, then you need to use something other than  `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`. Since those standard variable names are already in use for our 'main' credentials. There should be some way to pass other values in to the AWS SDK though.

File: lib/tasks/copy_database_to_s3.rake [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    Copy the latest copy of the database to a location on s3.
+
+    AWS_ACCESS_KEY_ID=joe AWS_SECRET_ACCESS_KEY=schmo bundle exec rake scihist:copy_database_to_s3
+    Note: this is a Postgres binary .dump file, not ASCII SQL. You can convert it to regular sql like this: pg_restore -f mydatabase.sql latest.dump
+  """
+
+  task :copy_database_to_s3 => :environment do
+    cmd = TTY::Command.new(output: Rails.logger)
+    get_database_url_args = ['heroku', 'pg:backups:url']
+    result =  cmd.run!(*get_database_url_args, binmode: true, only_output_on_error: true)
```
Comment (jrochkind) [inline lib/tasks/copy_database_to_s3.rake:15]: I don't think there's any benefit to abstract this as `get_database_url_args`, it is easier to read if you just put them right in there:

   cmd.run!("heroku", "pg:backups:url"....

**Comment by jrochkind** [comment]
Also I think I'm still confused about what we were talking about. I thought you were saying that you could provide a presigned S3 URL as an arg to `heroku pg:backups:url` or some similar command, so we wouldn't need to use down/put, we could just tell heroku where we wanted it to write backups too. But no?

If no, then this does look like the way to do it, and indeed this way doesn't use presigned URLs anymore, because there is no purpose for it, cool.

File: lib/tasks/copy_database_to_s3.rake (hunk 1 above)
Comment (jrochkind) [reply lib/tasks/copy_database_to_s3.rake:15]: I also don't think there's any reason this needs `binmode`?

**Comment by eddierubeiz** [comment]
Yeah, I realized after we talked that there is in fact no way to get Heroku to upload its own db backup to an arbitrary presigned S3 URL; I misinterpreted the docs for another function and realized belatedly that I'd been barking up the wrong tree. There's indeed no need for a presigned url!

**Comment by jrochkind** [comment]
> Yeah, I realized after we talked that there is in fact no way to get Heroku to upload 

Cool, the good news is that you still figured out a very simple way to do what's needed, great!

**Comment by jrochkind** [comment]
> Check that there is a DB to download before proceeding; fail gracefully if there isn't.

I don't think there's any need for that. We probably want it to fail hard with an error if that happens, it's not expected. 

Ideally that error would get caught/reported by honeybadger. So that's actually a checkbox to add, verify that honeybadger is going to catch errors reported by heroku scheduler regularly scheduled rake tasks, or figure out how to get them to be.

**Pull request merged**